04-onion-routing.md

```diff
@@ -254,14 +254,17 @@ The creator of `encrypted_recipient_data` (usually, the recipient of payment):
     - `total_fee_proportional_millionths(n+1) = ((total_fee_proportional_millionths(n) + fee_proportional_millionths(n+1)) * 1000000 + total_fee_proportional_millionths(n) * fee_proportional_millionths(n+1) + 1000000 - 1) / 1000000`
   - MUST create the `encrypted_recipient_data` from the `encrypted_data_tlv` as required in [Route Blinding](#route-blinding).
 
-The writer of `tlv_payload`:
+The writer of the TLV `payload`:
 
   - For every node inside a blinded route:
     - MUST include the `encrypted_recipient_data` provided by the recipient
     - For the first node in the blinded route:
       - MUST include the `blinding_point` provided by the recipient in `current_blinding_point`
     - If it is the final node:
       - MUST include `amt_to_forward`, `outgoing_cltv_value` and `total_amount_msat`.
+      - `outgoing_cltv_value` MUST be set to any additional CLTV delta which was added in addition
+        to the required amount for the full blinded path for privacy. This may be zero. This MUST
+        NOT include any additional amount added to compensate for a block during payment relay.
```

Comment:
While I understand what you mean because that's very close to what we implemented in eclair, I find this really hard to read, especially for new readers trying to understand blinded payments without a lot of implementation context. What about detailing more explicitly every step of the construction of `outgoing_cltv_value`, like this:

```
- When creating `outgoing_cltv_value`:
  - MUST use the current block height as a baseline
  - SHOULD add a random number of blocks for additional privacy
```

Why don't you want the sender to include an additional amount to compensate for a block found during payment relay? I don't see a good reason to prevent that, senders should be free to do so if they want to (especially since adding a random value to protect the sender privacy actually is also a way to compensate for blocks found during payment relay)?

Also, I think that what may be missing and is confusing for implementers is that recipients should include their `min_final_cltv_expiry_delta` in the `cltv-expiry-delta` for their blinded paths. That wasn't included here because I figured it was the responsibility of the Bolt 12 PR to add this since it's where we actually create blinded paths, but maybe it's worth adding here as well.

Reply:
Sorry for coming back late on this, I created the following gist to detail through examples how the spec currently expects nodes to behave: https://gist.github.com/t-bast/b1371d357a2c5f3e8c09514a62db7079

Eclair was not correctly implementing one of the final checks, which I'm fixing in https://github.com/ACINQ/eclair/pull/2678

I think this works, but this is quite subtle, please don't trust me on this and challenge it as much as possible!

Reply:
> Sorry for coming back late on this, I created the following gist to detail through examples how the spec currently expects nodes to behave: https://gist.github.com/t-bast/b1371d357a2c5f3e8c09514a62db7079
> 
> Eclair was not correctly implementing one of the final checks, which I'm fixing in [ACINQ/eclair#2678](https://github.com/ACINQ/eclair/pull/2678)
> 
> I think this works, but this is quite subtle, please don't trust me on this and challenge it as much as possible!

@t-bast Just took another look at this doc because we ran into some CLTV expiry issues. It may seem obvious, but FYI I think the spec is missing that the receiver needs to include their `min_final_cltv_expiry_delta` in the aggregated CLTV delta. In the example aggregation calculation in `proposals/route_blinding.md`, the receiver does not add their final delta. Let me know if I'm missing something here!

Reply:
Yes, that's correct! I'll prepare a PR to add this to the example in `proposals/route_blinding.md`. Note that the recipient may also use a dummy hop, in which case they don't necessarily need to add a `min_final_cltv_expiry_delta`, they would just cover it with the `cltv_expiry_delta` of the dummy hop.

Reply:
Done in https://github.com/lightning/bolts/pull/1131